docs/guides/5-custom-functions.md

```diff
@@ -287,6 +289,83 @@ As soon as you set a ruleset using `setRuleset()` or `loadRuleset()` method, eac
 
 You can store any kind of data, using cache for exchanging information between subsequent function calls is strongly discouraged. Also, Spectral does not guarantee any particular order of execution meaning the functions can be executed in random order, depending on the rules you have, and the document you lint.
 
+### Lifecycle events
+
+<!-- theme: info -->
+
+> This is a fairly advanced feature you should not need for most of the time.
+> Please do not use it unless you really do need to.
+
+Each custom function has access to `spectral` that, similarly to `fetch`, is exposed globally,
+At the moment, the instance has one method implemented, namely `on`.
+The method lets you subscribe to certain events that are emitted by Spectral during the linting process.
+At the time of writing this document, these events are:
+- setup - emitted before the linting process kicks in
+- beforeTeardown - after all rules have been executed (still possible to alter results at this point)
+- afterTeardown - linting done, no more actions possible.
+
+In practice, `setup` and `afterTeardown` should be used to clear any state, while `beforeTeardown` is the right even to perform any deferred operation.
+
+Example:
+
+```js
+let state = null;
+const _catch = () => null;
+
+spectral.on('setup', () => {
+  let resolve;
+  let reject;
+  const request = new Promise((_resolve, _reject) => {
+    resolve = _resolve;
+    reject = _reject;
+  })
+    .catch(_catch)
+
+  state = {
+    words: [],
+    request,
+    resolve,
+    reject,
+  };
+});
+
+spectral.on('beforeTeardown', () => {
+  if (state.words.length > 0) {
+    makeRequest(state.words)
+      .then(state.resolve)
+      .catch(state.reject);
+  } else {
+    state.reject();
+  }
+});
+
+spectral.on('afterTeardown', () => {
+  state = null;
+});
+
+module.exports = async function (targetVal) {
+  if (!state.words.includes(targetVal)) {
+    state.words.push(targetVal);
+  }
+
+  const dictionary = await state.request;
```

Comment:
...but it's used here (and thus seems pretty important). This portion seems to "automagically" trigger the `beforeTeardown` event. After careful reading it makes sense, but that seems very convoluted and not very explicit.

I imagine this has been designed that way so that the main function signature stays the same. But I wonder if a different approach may actually make it a bit easier to write and later reason about the code.

Idea:

Custom functions can either export a function OR an object.
- functions are used for simple and straightforward rules (the current implementation)
- objects should be preferred for deferred use cases

An possible object structure could be:
```
{
  "onSetup": // a sync/async function accepting no input parameter and returning no output. Called once by the runtime before the linting starts.
  "main": // the current sync/async function prototype with the following twist: it doesn't return any output. Called as many times as there are targetVals to analyse
  "onProduceResults": // a sync/async function accepting no input parameter and returning the list of results. Called once by the runtime once all targetVals have been visited.
  "onCleanup": // a sync/async function accepting no input parameter and returning no output. Called once by the runtime after the results have been retrieved.
}
```

I understand this might need a rework of the whole proposal, however I _think_ this would lead to easier to write code for the custom function author.

@P0lip Thoughts?

Reply:
> I imagine this has been designed that way so that the main function signature stays the same. But I wonder if a different approach may actually make it a bit easier to write and later reason about the code.

Yeah, FWIW, I wrote an internal email last week and proposed the following implementation.

```ts
module.exports = {
async setup() {
   // executed before linting is performed
},

async run() {
 // this would be called in the same manner as current custom fn is
},

async teardown() {
   // executed after linting is performed
  // you can still manipulate the results at this point
}
};
```
This pretty much aligns with your proposal - the naming is simply tad different, but the concept is fairly similar.
I didn't implement it the exact way eventually, as I wanted to avoid changing the signature.
It'd just add more burden to our code, while the usage of it would be rather marginal.

> I understand this might need a rework of the whole proposal, however I think this would lead to easier to write code for the custom function author.

To be frank, it wouldn't be that much of work _if_ we changed it to:
```
spectral.onSetup
spectral.onProduceResults
spectral.onCleanup
spectral.onLintTarget
// no function exported **at all**
```
That's not exactly the same, however, and it may look more complicated.

The primary reason I decided to inject this `spectral` object is that it may be useful for us in future at some point.